fix(docs-infra): do not redirect docs URLs on archive deployments

gkalpak opened the pull request:
To avoid showing outdated info (such as events, resources, etc.) but still allow people to see docs for older versions, we redirect non-documentation URLs to `/docs`. Recently(-ish) we have added documentation content under the `/cli/...` and `/start/...` path-prefixes, but we haven't added them to the list of documentation URLs that should not be redirected. As a result, on archive deployments (e.g. https://v7.angular.io/cli), they are redirected to `/docs`, making it impossible to see the documentation for these versions (unless you know about the `?mode=stable` work-around).

This commit fixes it by adding `cli` and `start` to the list of documentation URLs that are excluded from redirection.

##
Current incorrect behavior (redirects to `/docs`): https://v7.angular.io/cli
New correct behavior (stays on `/cli`): https://pr30894-0c145e8.ngbuilds.io/cli?mode=archive

gkalpak commented:
**merge-assistance:** If possible, we need this in master, 8.x and 7.x.

mary-poppins commented:
You can preview 0c145e8 at https://pr30894-0c145e8.ngbuilds.io/.

petebacondarwin reviewed (commented):
Perhaps we should flip the logic here and redirect paths (other than `docs`) that have only one segment... e.g.

```ts
if (this.deployment.mode === 'archive' &&
    path !== 'docs' &&
    path.indexOf('/') === -1) {
  ...
```

gkalpak commented:
There are several one-segment paths that are docs: `docs`, `cli`, `start`, `tutorial`

petebacondarwin reviewed (approved):
@gkalpak - good point! But I feel there should be a less hard-coded approach. We are able to determine marketing docs another way, since we are able to provide a different layout for them (i.e. no left panel).  This is the code that decides that:

```ts
    const isSideNavDoc = !!this.currentNodes[sideNavView];
```

Perhaps we can do something along those lines?

Otherwise happy with this PR as it is.

gkalpak committed:
refactor(docs-infra): make archive redirection tests DRY

gkalpak committed:
fix(docs-infra): do not redirect docs URLs on archive deployments

To avoid showing outdated info (such as events, resources, etc.) but
still allow people to see docs for older versions, we redirect
non-documentation URLs to `/docs`. Recently(-ish) we have added
documentation content under the `/cli/...` and `/start/...`
path-prefixes, but we haven't added them to the list of documentation
URLs that should not be redirected. As a result, on archive deployments
(e.g. https://v7.angular.io/cli), they are redirected to `/docs`, making
it impossible to see the documentation for these versions (unless you
know about the `?mode=stable` work-around).

This commit fixes it by adding `cli` and `start` to the list of
documentation URLs that are excluded from redirection.

gkalpak committed:
refactor(docs-infra): avoid hard-coding URLs to redirect on archive mode

Related discussion:
https://github.com/angular/angular/pull/30894#pullrequestreview-246731995

mary-poppins commented:
You can preview e2a6876 at https://pr30894-e2a6876.ngbuilds.io/.

gkalpak commented:
Added a couple of refactoring commits (one of which switches to using `currentNodes[sideNavView]`.
WDYT, @petebacondarwin?

petebacondarwin commented on aio/src/app/app.component.spec.ts:
```diff
@@ -809,106 +810,44 @@ describe('AppComponent', () => {
+      Object.keys(redirectionPerMode).forEach(mode => {
+        const doRedirect = redirectionPerMode[mode];
+        const description =
+            `should ${doRedirect ? '' : 'not '}redirect to 'docs' if deployment mode is '${mode}' ` +
```
Not using the backspace char? :-P

gkalpak replied:
I would, but it doesn't display well in the browser (with our [kjhtml reporter][1]) :grin:

[1]: https://www.npmjs.com/package/karma-jasmine-html-reporter

petebacondarwin commented on aio/src/app/app.component.spec.ts:
```diff
@@ -809,106 +810,44 @@ describe('AppComponent', () => {
+        const doRedirect = redirectionPerMode[mode];
+        const description =
+            `should ${doRedirect ? '' : 'not '}redirect to 'docs' if deployment mode is '${mode}' ` +
+            'and not at a docs page';
```
`not at a docs page` -> `at a marketing page`?

gkalpak replied:
I was keeping what was there before, but happy to go with `marketing page` :+1:

petebacondarwin commented on aio/src/app/app.component.spec.ts:
```diff
@@ -809,106 +810,44 @@ describe('AppComponent', () => {
+        const verifyNoRedirection = () => expect(TestBed.get(LocationService).replace).not.toHaveBeenCalled();
+        const verifyPossibleRedirection = doRedirect
+            ? () => expect(TestBed.get(LocationService).replace).toHaveBeenCalledWith('docs')
+            : verifyNoRedirection;
```
Why not make:

```
const verifyRedirection = () => expect(TestBed.get(LocationService).replace).toHaveBeenCalledWith('docs');
```

And then `verifyPossibleRedirection` becomes super clean:

```
const verifyPossibleRedirection = doRedirect ? verifyRedirection : verifyNoRedirection;
```

petebacondarwin reviewed (approved):
Love it. A couple of nits...

gkalpak committed:
fixup! refactor(docs-infra): make archive redirection tests DRY

mary-poppins commented:
You can preview 8ee6437 at https://pr30894-8ee6437.ngbuilds.io/.
